### PR TITLE
Add --file-locks checkpoint/restore option

### PR DESCRIPTION
**Pull request by rst0git**
#### What this PR does / why we need it:

CRIU supports checkpoint/restore of file locks. This feature is required to checkpoint/restore containers running applications such MySQL.

#### How to verify it?

```bash
sudo podman run --name some-mysql -e MYSQL_ROOT_PASSWORD=my-secret-pw -d mysql:latest

sudo podman container checkpoint --file-locks some-mysql
```

### CODE REVIEW

**Comment by TomSweeneyRedHat** [comment]
Quick review LGTM, but you'll need to update the man pages with the new option before the CI will let this get through.

File: test/e2e/checkpoint_test.go
Comment (edsantiago) [inline test/e2e/checkpoint_test.go:1355]: Does this need a custom crun? It just fails for me with:
```console
$ sudo bin/podman container checkpoint 3a62 --file-locks
checkpoint: unrecognized option '--file-locks'
Try `checkpoint --help' or `checkpoint --usage' for more information.
Error: `/usr/bin/crun checkpoint --image-path /var/lib/containers/storage/overlay-containers/3a62ce7ff8f88b739d192eafedea5717cb224e1bfac8e573ea3ed85dd5ce468e/userdata/checkpoint --work-path /var/lib/containers/storage/overlay-containers/3a62ce7ff8f88b739d192eafedea5717cb224e1bfac8e573ea3ed85dd5ce468e/userdata --file-locks 3a62ce7ff8f88b739d192eafedea5717cb224e1bfac8e573ea3ed85dd5ce468e` failed: exit status 64
```

**Review by edsantiago — CHANGES_REQUESTED** [review]
Very quick review: you need to remove `-l` and resubmit (or, alternately, skip if remote).

File: test/e2e/checkpoint_test.go
Comment (rst0git) [reply test/e2e/checkpoint_test.go:1355]: It should work with the patch in https://github.com/containers/crun/pull/783, but it hasn't been released yet.

Comment (edsantiago) [reply test/e2e/checkpoint_test.go:1355]: Then there's not much likelihood that CI will pass. (Current failures seem to be a registry outage. I'm not going to hit Re-run since there's no point).

File: test/e2e/checkpoint_test.go
Comment (rst0git) [reply test/e2e/checkpoint_test.go]: Thanks, I replaced `-l` with container name.

File: test/e2e/checkpoint_test.go
Comment (rst0git) [reply test/e2e/checkpoint_test.go]: I've added a condition to run the test only with `runc` and a FIXME comment to enable the test for `crun`.

File: test/e2e/checkpoint_test.go
Comment (edsantiago) [inline test/e2e/checkpoint_test.go:1346]: Could you change this to `Skip("FIXME: requires crun >= 1.4")`? I have a `find-obsolete-skips` script that I run from time to time, and a clear FIXME would make it easy for me to remove this in the next few months.

File: test/e2e/checkpoint_test.go
Comment (edsantiago) [inline test/e2e/checkpoint_test.go:1356]: I'm fond of checking error messages, not just exit code. (The command could fail for other reasons, even a typo). Unfortunately criu does not emit useful diagnostics, but even just checking for the "blah blah -52" message would be better than nothing.

**Review by edsantiago — COMMENTED** [review]
Some minor suggestions (on the tests only; I don't feel qualified to comment on the rest).

Please hold off on re-pushing: all of CI is currently broken. There's a fix in progress, but you may need to rebase on main once that's in. Sorry for the inconvenience.

File: test/e2e/checkpoint_test.go
Comment (rst0git) [reply test/e2e/checkpoint_test.go]: Thanks, I've updated the skip message.

File: test/e2e/checkpoint_test.go
Comment (rst0git) [reply test/e2e/checkpoint_test.go:1356]: Thanks, I've added a check for the error message.

**Review by edsantiago — APPROVED** [review]

**Comment by edsantiago** [comment]
Flake is a combination of #12273 and #12306. Restarting.

**Comment by edsantiago** [comment]
Other CI failures are [real](https://storage.googleapis.com/cirrus-ci-6707778565701632-fcae48/artifacts/containers/podman/6360175809396736/html/int-podman-fedora-34-root-container.log.html#t--podman-checkpoint-a-running-container-by-id--1):
```
Running: podman [options] container restore e6b6a148e5d9abfe053ce8bc0d93d3e2fcef1790e4d4f2a7773ab691d65609f7
Error: OCI runtime error: restore: unrecognized option '--file-locks'
```

**Comment by rst0git** [comment]
> Other CI failures are [real](https://storage.googleapis.com/cirrus-ci-6707778565701632-fcae48/artifacts/containers/podman/6360175809396736/html/int-podman-fedora-34-root-container.log.html#t--podman-checkpoint-a-running-container-by-id--1):
> 
> ```
> Running: podman [options] container restore e6b6a148e5d9abfe053ce8bc0d93d3e2fcef1790e4d4f2a7773ab691d65609f7
> Error: OCI runtime error: restore: unrecognized option '--file-locks'
> ```

Thanks, this should be fixed now.

**Comment by edsantiago** [comment]
LGTM. @containers/podman-maintainers PTAL

**Comment by umohnani8** [comment]
LGTM

**Comment by edsantiago** [comment]
Aw, phooey. @rst0git there's a merge conflict, looks like #12342

**Comment by rst0git** [comment]
> Aw, phooey. @rst0git there's a merge conflict, looks like #12342

Thanks, I've rebased the branch.

**Comment by mheon** [comment]
@adrianreber Mind doing a quick review?

Code LGTM

**Comment by rhatdan** [comment]
/approve
/lgtm
Thanks @rst0git

**Comment by adrianreber** [comment]
> @adrianreber Mind doing a quick review?
> 
> Code LGTM

I am too late, but looks good.  One thing, however, shouldn't there be a check if the runtime actually supports `--file-locks`? There is a PR referenced for crun 1.4, but currently it will fail with crun. Although I think there are a couple of things (pre-dump for example) were we do not have support in crun but only in runc.

**Comment by rst0git** [comment]
> shouldn't there be a check if the runtime actually supports --file-locks?

We could check what is the runtime and its version and if it is an older version of `crun` we could:
1. Display an error message to update `crun` and/or mention that users could add `file-locks` in `/etc/criu/default.conf`;
2. Or generate a config file for CRIU and specify the option in this way.

> I think there are a couple of things (pre-dump for example)

For `--pre-checkpoint` we could display appropriate error message for now, and I will look into adding pre-dump support in `crun`.

@adrianreber What do you think?

**Comment by rhatdan** [comment]
I am not crazy about adding assumptions like that to the code, since over time they will just become stale.  That should probably be a distributions problem, in that it should ensure the OCI runtimes are new enough.

**Pull request merged**
